contracts/src/utils/cryptography/ecdsa.rs

```diff
@@ -0,0 +1,248 @@
+///
+/// # Arguments
+///
+/// * `storage` - Write access to storage.
+/// * `hash` - Hash of the message.
+/// * `v` - `v` value from the signature.
+/// * `r` - `r` value from the signature.
+/// * `s` - `s` value from the signature.
+///
+/// # Errors
+///
+/// * If the `s` value is grater than `EIP2_VALUE`, then the error
+/// [`Error::InvalidSignatureS`] is returned.
+/// * If the recovered address is `Address::ZERO`, then the error
+/// [`Error::InvalidSignature`] is returned.
+///
+/// # Panics
+///
+/// * If `ecrecover` precompile fail to execute.
+pub fn recover(
+    storage: &mut impl TopLevelStorage,
+    hash: B256,
+    v: u8,
+    r: B256,
+    s: B256,
+) -> Result<Address, Error> {
+    validate_s_value(&s)?;
+    // If the signature is valid (and not malleable), return the signer address.
+    evm_recover(storage, hash, v, r, s)
+}
+
+/// Calls `ecrecover` EVM precompile.
+/// The `ecrecover` EVM precompile allows for malleable (non-unique) signatures:
+/// this function rejects them by requiring the `s` value to be in the lower
+/// half order, and the `v` value to be either 27 or 28.
+///
+/// # Arguments
+///
+/// * `storage` - Write access to storage.
+/// * `hash` - Hash of the message.
+/// * `v` - `v` value from the signature.
+/// * `r` - `r` value from the signature.
+/// * `s` - `s` value from the signature.
+///
+/// # Errors
+///
+/// * If the `s` value is grater than `EIP2_VALUE`, then the error
+/// [`Error::ECDSAInvalidSignatureS`] is returned.
+/// * If the recovered address is `Address::ZERO`, then the error
+/// [`Error::InvalidSignature`] is returned.
+///
+/// # Panics
+///
+/// * If `ecrecover` precompile fail to execute.
+fn evm_recover(
+    storage: &mut impl TopLevelStorage,
+    hash: B256,
+    v: u8,
+    r: B256,
+    s: B256,
+) -> Result<Address, Error> {
+    let calldata = prepare_calldata(hash, v, r, s);
+
+    if v == 0 || v == 1 {
+        // `ecrecover` panics for this values
+        // but following the Solidity tests
+        // https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/test/utils/cryptography/ECDSA.test.js
+        // it should return `ECDSAInvalidSignature` error.
+        return Err(ECDSAInvalidSignature {}.into());
+    }
+
+    let recovered =
+        call::static_call(Call::new_in(storage), ECRECOVER_ADDR, &calldata)
+            .expect("should call `ecrecover` precompile");
+
+    let recovered = B256::from_slice(&recovered);
+    let recovered = Address::from_word(recovered);
+
+    if recovered.is_zero() {
+        return Err(ECDSAInvalidSignature {}.into());
+    }
+    Ok(recovered)
+}
+
+/// Prepares call data for `ecrecover` EVM precompile.
+///
+/// # Arguments
+///
+/// * `hash` - Hash of the message.
+/// * `v` - `v` value from the signature.
+/// * `r` - `r` value from the signature.
+/// * `s` - `s` value from the signature.
+fn prepare_calldata(hash: B256, v: u8, r: B256, s: B256) -> Vec<u8> {
```

Comment:
nit: maybe call this `abi_encode`? maybe `encode_calldata`? `prepare` is not a very useful word.